sql/2024/fonts/development/fonts_color_color.sql

```diff
@@ -0,0 +1,63 @@
+-- Section: Development
+-- Question: What is the distribution of color palettes?
+-- Normalization: Fonts (color only)
+
+-- INCLUDE ../common.sql
+
+CREATE TEMPORARY FUNCTION COLORS(json STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js
+AS '''
+function toHex(value) {
+  return ('0' + (value & 0xFF).toString(16)).slice(-2);
+}
+
+try {
+  const $ = JSON.parse(json);
+  const result = new Set();
+  for (const palette of $) {
+    for (const [blue, green, red, alpha] of palette) {
+      result.add(`#${toHex(red)}${toHex(green)}${toHex(blue)}${toHex(alpha)}`);
+    }
+  }
+  return Array.from(result);
+} catch (e) {
+  return [];
+}
+''';
+
+WITH
+fonts AS (
+  SELECT
+    client,
+    url,
+    COLORS(JSON_EXTRACT(ANY_VALUE(payload), '$._font_details.color.palettes')) AS colors,
```

Comment:
Why ANY_VALUE?

Reply:
So we group by the client and URL. For that combination, there are typically many entries corresponding to requests performed from different pages and at different times. It is then assumed the server's response does not depend on the location and time, and we just take any one here and in all other queries with the same grouping.

Reply:
Oh OK. I expected to look at fonts per site, but here we're looking at fonts overall (no matter which site requests them). OK.